controllers/onload_controller.go

```diff
@@ -56,15 +73,197 @@ func (r *OnloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 	// indicated by the deletion timestamp being set.
 	isOnloadMarkedToBeDeleted := onload.GetDeletionTimestamp() != nil
 
-	if !isOnloadMarkedToBeDeleted {
-		log.Info("Adding new Onload kind")
-	} else {
-		log.Info("Deleting new Onload kind")
+	if isOnloadMarkedToBeDeleted {
+		ok, res, err := r.deleteModules(ctx, onload)
+		if !ok {
+			return res, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("Adding new Onload kind", "onload.Name",
+		onload.Name, "onload.Namespace", onload.Namespace)
+
+	ok, res, err := r.createAndAddModules(ctx, onload)
+	if !ok {
+		return res, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *OnloadReconciler) deleteModules(
+	ctx context.Context,
+	onload *onloadv1alpha1.Onload,
+) (bool, ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	module := &kmm.Module{}
+	err := r.Get(
+		ctx,
+		types.NamespacedName{Name: onload.Name, Namespace: onload.Namespace},
+		module,
+	)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "Failed to get onload-module")
+		return false, ctrl.Result{}, err
+	}
+	if err == nil {
+		// I couldn't find a way to explicitly determine whether the module
+		// was in the process of deletion, the best I found was to compare
+		// the desired number of pods against the the actual number
+		if module.Status.ModuleLoader.AvailableNumber >
+			module.Status.ModuleLoader.DesiredNumber {
```

Comment:
Does this check exist, so we don't call the `Delete()` with the module twice? I've started looking at the Conditions https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go, and it looks like a tool for this situation:
1. Create the "delete pending" condition as "true" after `Delete()`.
2. Check that condition on the next reconciliation.
3. If the condition is "true", but the module is not found, set it to "false".
4. Make further progression with the deletion.

I need to poke its API to understand further.

Reply:
I don't believe it's necessary (or even possible, given the near infinite possible failure conditions) to diagnose external state. Such messages may be well intentioned but misleading.

If the cluster isn't as desired, just call the API (eg. delete CR) and carry on if `err == nil`. When the next step (eg. create CR with same kind+name) fails, just fail it and let the reconciler start again.

Reply:
I'm suggesting a general philosophy of simply fail or continue, where [continuing](https://github.com/rh-ecosystem-edge/kernel-module-management//blob/main/controllers/module_reconciler.go#L143C6-L143C6) means letting other reconciliation actions get started and tackle the action again on next loop. I'm suggesting in distributed computing it's impossible for one process to truly know the reason for an incomplete cluster action. For example, failure may be nothing to do with normal operation of the cluster and actually due to a temporary inconsistency in the distributed database data or network/storage hardware issue.

With KMM Hub, which creates a `ManifestWork` (essentially an exact equiv of our Onload CR as [it](https://kmm.sigs.k8s.io/documentation/hub_spoke/) "contains a trimmed-down `Module` resource"), they use [createOrPatch](https://github.com/rh-ecosystem-edge/kernel-module-management//blob/main/controllers/hub/managedclustermodule_reconciler.go#L135) when deletion is not necessary and [simply fail](https://github.com/rh-ecosystem-edge/kernel-module-management/blob/main/internal/manifestwork/manifestwork.go#L112) upon a failed deletion when clearing up.